reps/2022-09-19-ray-on-spark.md

```diff
@@ -0,0 +1,160 @@
+## Summary
+### General Motivation
+
+Supporting running Ray applications on spark cluster / databricks runtime,
+Spark is a popular distributed computing framework and is used widely,
+If supporting running Ray applications on spark cluster, user don't need to
+setup a standalone ray cluster and it allows ray workloads and spark workloads
+runs together.
+
+#### Key requirements:
+- Ray application resources (cpu/gpu/memory) allocation must respect spark job resource allocation.
+- Little overhead compared with native Ray cluster.
+- (Optional) Ray UI portal support.
+
+### Should this change be within `ray` or outside?
+
+Yes. For better code maintemance.
+
+## Stewardship
+### Required Reviewers
+
+TODO
+
+### Shepherd of the Proposal (should be a senior committer)
+
+TODO
+
+## Design and Architecture
+
+### Preparation
+
+We need to have a spark cluster (apache/spark >= 3.3)
+We need install ray packages on every node of the cluster, by:
+```
+pip install ray
+pip install "ray[debug,dashboard,tune,rllib,serve]"
+```
+
+### How to setup Ray cluster over spark cluster ?
+
+A spark cluster is like:
+![spark-cluster-overview](https://spark.apache.org/docs/latest/img/cluster-overview.png)
+
+Setup ray cluster over spark cluster, we can do:
+- Create a spark barrier mode job,  a spark barrier mode job means all tasks in this spark job 
+will be executed concurrently, and each task is allocated with several resources, by default,
+one task is allocate with 1 cpu core. If we want to setup a Ray cluster with  16 cpu cores resources
+in total, we can create a spark barrier mode with 16 tasks.
+ 
+- Inside each spark executor, launch one Ray Node (head / worker) and keep the spark job running until
+the Ray cluster destroyed. Note: one spark executor might contain multiple spark tasks,
+but we will only launch one ray node for each spark executor.
+e.g., suppose in executor1, there’re 3 spark tasks launched, then in task of local-rank-0,
+we launch a ray node with –num-cpus=3. Using the following commands to launch ray node in spark tasks:
+  - **Head node:** ray start —head --num-cpus=X --num-gpus=Y --memory=Z
+  - **Worker node:** ray start —head --num-cpus=X --num-gpus=Y --memory=Z --address={head_ip}:{port}
```

Comment:
This part is a little strange, it seems Ray nodes end up with potentially different / random CPUs assigned depending on how Spark happened to schedule the tasks. In many use cases, users may want to have nodes with specific numbers of resources (e.g., 8-CPU nodes, or 4-GPU nodes).

Would the API allow for configuration of that kind of cluster?

Reply:
>  it seems Ray nodes end up with potentially different / random CPUs assigned depending on how Spark happened to schedule the tasks. 

No. The resources amount allocated is deterministic.

e.g., your case: (8-CPU nodes, or 4-GPU nodes), 
suppose on a spark cluster with config:
```
spark.task.cpus 2 # means 2 cpu cores per spark task
spark.task.resource.gpu.amount 1 # means 1 GPU per spark task
```

Then the Ray on spark routine will create a spark job with **4 spark tasks** running concurrently, which means it books (8-CPU nodes, or 4-GPU nodes) resources for ray cluster.


One issue is the CPU / GPU cores ratio depends on spark cluster config, this is a limitation.

Reply:
Cluster level total resources is not enough. We need to control at the per node level. Otherwise, some tasks may not be able to run if there is no single node that has the enough resources.

Reply:
Got it.
What's the minimal resources requirements for a Ray Node ?

Reply:
This might be a limitation on spark cluster,

We cannot control the spark task distribution on spark cluster, it is possible that one executor only launch one spark task, in this case, the corresponding Ray Node will get resources amount equal to that one spark task gets.

Reply:
But, we can increase per-spark-task resources config , by increase config of:
```
spark.task.cpus
spark.task.resource.gpu.amount
``

Reply:
Setting spark.task.cpus/gpus sounds like the way to go. It sounds like this can be set dynamically per Spark job? That's probably the ideal case, though it would limit you to Ray clusters of homogeneous node shape.

The API could look like `ray.spark.init_cluster(num_nodes=100, num_cpus_per_node=16, num_gpus_per_node=4)`

Reply:
> It sounds like this can be set dynamically per Spark job?

Unfortunately this is spark cluster level setting, we cannot set it dynamically per spark job.

Reply:
> Ray clusters of homogeneous node shape

I have a question @ericl 

For a ray cluster, suppose on a worker machine, we launch multiple ray node instances, shall we merge it into one ray node ? 
I prefer to merge it into one ray node so that it is seemingly beneficial for performance.

Reply:
Another issue is, because we setup ray cluster on spark cluster, for each spark job, it's spark tasks distribution might be imbalanced, e.g.

A spark cluster with 2 worker machine (worker_1 and worker_2), each worker machine has 4 cpu cores. The spark cluster is shared with user A and user B.

Now, suppose user_A launch a spark job-A, job_A generates 4 spark tasks, 3 spark tasks launched in worker_1, and 1 spark task launched in worker_2.

At the same time, user_B launch a second spark job_B (job B is used to run a ray cluster), then the job_B generates 1 spark tasks  in worker_1, 3 spark tasks in worker_2.

For job_B, the ray cluster, we can get 1 cpu cores in worker_1, and 3 cpu cores in worker_2,
in my proposal ray on spark design, I will start a ray worker node in worker_1 with num-cpus=1 , and start a ray worker node in worker_2 with num-cpus=3  @ericl WDYT ?

Or you think we should start 3 ray worker nodes (each with num-cpus=1) in worker_2 ?

Reply:
> What's the minimal resources requirements for a Ray Node ?

Typically we don't recommend fewer than 8 CPUs / 30GB per node. Maybe 4 CPUs is a reasonable minimum.

> For a ray cluster, suppose on a worker machine, we launch multiple ray node instances, shall we merge them into one ray node ?

I think this would be generally preferable, yes, since Ray works better with fewer large nodes.

> in my proposal of ray on spark design, it will start a ray worker node in worker_1 with num-cpus=1 , and start a ray worker node in worker_2 with num-cpus=3 @ericl WDYT ?

It could work, but I would see it causing unexpected issues. For example, it could accidentally create Ray clusters with many tiny nodes, or sometimes few big nodes, even when running on the same Spark cluster, just due to how other workloads happen to be occupying executors. This would make the performance / reliability of Ray-on-Spark workloads unpredictable in shared clusters.

It seems like it would be preferable to only support non-shared clusters, if it could guarantee more consistently predictable cluster setups.

Btw, there's not just cpus to be allocated; there's object_store_memory, which is typically set to ~30% of the node space. You'd need to make sure this is set proportionally as well to avoid over-loading the node.

Reply:
@ericl 

> It seems like it would be preferable to only support non-shared clusters

Customer's spark cluster might also running other workloads, this is common cases, so we should support shared clusters as well.

> it could accidentally create Ray clusters with many tiny nodes, or sometimes few big nodes

For preventing tiny nodes, we can increase configs `spark.task.cpus` and `spark.task.resource.gpu.amount`, so that it can set the lower bound of the node shape.

For preventing big nodes, we can split big ray node into multiple ray nodes.

> Btw, there's not just cpus to be allocated; there's object_store_memory, which is typically set to ~30% of the node space

When increasing `spark.task.cpus` and `spark.task.resource.gpu.amount`, each spark worker generates less spark tasks slots, it means each task can be allocated with more memories. We can provide helper method to tell user the optimal `spark.task.cpus` and `spark.task.resource.gpu.amount` setting satisfying Ray node requirements according to the spark worker memory capacity.

Reply:
Requiring the user to increase the task config size can work. I think it would be best to require the user to set these values to configure the Ray node size, and then always launch one node per task. Maybe we can also require the node to have at least 4 CPUs or raise an error.

Otherwise, the Ray cluster setup will be nondeterministic, and you could get very strange results with bad luck on the node sizing.

Reply:
> Otherwise, the Ray cluster setup will be nondeterministic, and you could get very strange results with bad luck on the node sizing

Do you mean different shape Ray worker nodes causing this ? If so, we can launch one ray node per spark task to prevent the issue.

Reply:
@ericl 

Btw, I will set Ray head node with `--num-cpus=0` (prevent ray task scheduled to head node), and then set all ray worker nodes with the same shape. Does the head node different shape have some influence ?

Reply:
That makes sense if you always put the head node on the Spark driver node. If the head node can live on other nodes, then I would consider setting num-cpus to be the same as the other nodes.